### DISCUSSION TITLE
Conditionally allow params on generic parameters if generic type is params eligible

### DISCUSSION THREAD

**Discussion by ThadHouse** [dotnet/csharplang#7840] [Language Ideas]
Was playing around with some generic code that could take either an array or a base type, and something I was thinking would be cool would be to allow `params` on a non array generic parameter, and only allow the caller to use params if the target type allows params.

Something like the following example.

```
public class Cup<T> {
    public void Add(params T value) { ... }
}
```

And then you could use this like the following.

```
Cup<string[]> c = new();
c.Add("a", "b", "c");
```

Whereas if T wasn't allowed to be params, the call would still work, but just wouldn't work like an array

```
Cup<string> c = new();
c.Add("a"); // allowed
c.Add("a", "b"); // errors with no matching overload
```

I know a lot of work to params is being looked at, so maybe this would be cool too.

**Comment by sharpchen** [comment]
Why not just add an overload, single parameter version has higher priority
```cs
new Cup<int>().Add(1); // void Cup<T>.Add(T value)

public class Cup<T> {
    public void Add(params T[] value) { Console.Write("from params"); }
    public void Add(T value) { Console.Write("from single"); }
}
```

**Reply by huoyaoyuan** [reply]
`params` is going to be extended on more types. It will be impractical to enumerate every supported type.

**Reply by sharpchen** [reply]
That can be a new feature, something like `params [T]` that supports all available collection-like types. But I think we are able to implement this with source generator, right?

**Reply by KennethHoff** [reply]
Collection-like types are just things that implement `IEnumerable` though, right? (... Assuming `ref struct` can implement interfaces in a future lang version, and they add it to Span and ROS).

**Reply by ThadHouse** [reply]
Just adding an array params overload doesn’t work, because it will do a different thing. It’s not taking multiple elements to add, it’s adding something that takes multiple elements. 

Think of adding it to List<T>. params T[] Add makes no sense, that’s just AddRange. But params T if T is an array does make sense, as it would make an array, and add that to the list.
